affinity/cpp-23/d1795r1.md

```diff
@@ -124,11 +132,135 @@ As a result of the above this paper may also:
 * Propose a lifetime model for execution agents.
 * Propose some additions to the C\+\+ machine model to facilitate describing these additional properties.
 
-# Suggested straw polls
+# 6. Proposal
+
+## Header `<system>` synopsis
+
+```cpp
+namespace std {
+namespace experimental {
+
+/* system_topology */
+
+class system_topology {
+
+  system_topology() = delete;
+
+  std::chrono::time_point<std::chrono::system_clock> timestamp() const noexcept;
+
+};
+
+/* system_resource */
+
+class system_resource {
+
+  system_resource() = delete;
+
+};
+
+/* traverse_topology */
+
+template <class T>
+std::vector<system_resource> traverse_topology(const system_topology &, const T &) noexcept;
+
+/* this_system::discover_topology */
+
+namespace this_system {
+
+system_topology discover_topology();
+
+}  // namespace this_system
+
+}  // experimental
+}  // std
+```
+
+## Terms of art
+
+The term *system resource* refers to a hardware or software abstraction of an execution, memory, network or I/O resource within a system.
+
+The term *system topology* refers to a possibly cyclic graph of *execution resources* connected to the abstract machine, and their various properties.
+
+> [*Note:* The current definition of *system topology* is currently incomplete and will be developed over the course of this proposal as the various C\+\+ domains are represented. *--end note*]
+
+The term *topology traversal policy* refers to a policy that describes the way in which a *system topology* is traversed in order to to produce a collection of *system resources*.
+
+## Class `system_topology`
+
+The `system_topology` class provides an abstraction of a read-only snapshot of the *system topology* at a particular point in time. A `system_topology` object may not maintain or otherwise be associated with the lifetime of operating system or third party library resources.
+
+### `system_topology` constructors
+
+```cpp
+system_topology() = delete;
+```
+
+*Effects:* Explicitly deleted.
+
+### `timestamp` member function
+
+```cpp
+std::chrono::time_point<std::chrono::system_clock> timestamp() const noexcept;
+```
+
+*Returns:* A `std::chrono::time_point<std::chrono::system_clock>` object representing the time at which the runtime discovery of the system topology performed to construct the `system_topology` object was completed.
```

Comment:
You make a good point here. My initial thinking was that discover_topology would perform topology discovery for the entire system and then if something changed you could then discover again and compare the new topology with the previous. But perhaps we do want the interface to be more fine grained than that.

I can take the timestamp member function out until we decide how granular the interface should be.

Reply:
@AerialMantis I think it's a good idea to take it out, if that's OK with you.  My worry is that there's no good way to define a single timestamp for a whole topology, since discovery is not atomic.  Providing a timestamp might give users the false impression that it is atomic.

Reply:
I agree FWIW, it's something the user can do if they want to that the library can't really do a better job of than they can.  It also might incur costs that would otherwise be unnecessary if we ever get to looking at compile-time visible versions or subsets.

Reply:
These are good points, I agree we should allow the user flexibility to choose when they discover different parts of the topology and timestamp in a way that is appropriate to their use case.

I have removed the `timestamp` member function from this revision.